⭐ feat(runner): More runner improvements

DatCaptainHorse opened the pull request:
## Description
Whew..

- Steam can now run without namespaces using live-patcher (because Docker..)
- Improved NVIDIA GPU selection and handling
- Pipeline tests for GPU picking logic
- Optimizations and cleanup all around
- SSH (by default disabled) for easier instance debugging.
- CachyOS' Proton because that works without namespaces (couldn't figure out how to enable automatically in Steam yet..)
- Package updates and partial removal of futures (libp2p is going to switch to Tokio in next release hopefully)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - SSH server can now be enabled within the container for remote access when configured.
  - Added persistent live patching for Steam runtime entrypoints to improve compatibility with namespace-less applications.
  - Enhanced GPU selection with multi-GPU support and PCI bus ID matching for improved hardware compatibility.
  - Improved encoder selection by runtime testing of video encoders for better reliability.
  - Added WebSocket transport support in peer-to-peer networking.
  - Added flexible compositor and application launching with configurable commands and improved socket handling.

- **Bug Fixes**
  - Addressed NVIDIA-specific GStreamer issues by setting new environment variables.
  - Improved error handling and logging for GPU and encoder selection.
  - Fixed process monitoring to handle patcher restarts and added cleanup logic.
  - Added GStreamer cache clearing workaround for Wayland socket failures.

- **Improvements**
  - Real-time logging of container processes to standard output and error for easier monitoring.
  - Enhanced process management and reduced CPU usage in protocol handling loops.
  - Updated dependency versions for greater stability and feature support.
  - Improved audio capture defaults and expanded audio pipeline support.
  - Enhanced video pipeline setup with conditional handling for different encoder APIs and DMA-BUF support.
  - Refined concurrency and lifecycle management in protocol messaging for increased robustness.
  - Consistent namespace usage and updated crate references across the codebase.
  - Enhanced SSH configuration with key management, port customization, and startup verification.
  - Improved GPU and video encoder integration in pipeline construction.
  - Simplified error handling and consolidated write operations in protocol streams.
  - Removed Ludusavi installation from container image and updated package installations.

- **Other**
  - Minor formatting and style changes for better code readability and maintainability.
  - Docker build context now ignores `.idea` directory to streamline builds.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DatCaptainHorse committed:
More runner improvements

- Steam can now run without namespaces using live-patcher (because Docker..)
- Improved NVIDIA GPU selection and handling
- Pipeline tests for GPU picking logic
- Optimizations and cleanup all around..

DatCaptainHorse commented on packages/server/src/p2p/p2p_protocol_stream.rs:
```diff
@@ -56,21 +59,35 @@ impl NestriStreamProtocol {
         Ok(sp)
     }
 
+    pub fn restart(&mut self) -> Result<(), Box<dyn std::error::Error>> {
+        // Return if tx and handles are already initialized
+        if self.tx.is_some() && self.read_handle.is_some() && self.write_handle.is_some() {
+            tracing::warn!("NestriStreamProtocol is already running, restart skipped");
+            return Ok(());
+        }
+
+        let (tx, rx) = mpsc::channel(1000);
+        self.tx = Some(tx);
+        self.read_handle = Some(self.spawn_read_loop());
+        self.write_handle = Some(self.spawn_write_loop(rx));
+
+        Ok(())
+    }
```
Yeah we aren't using it to actually restart yet, just first start.. so it's a future thing to deal with (in another PR). Resolving for now.

DatCaptainHorse commented on packages/server/src/main.rs:
```diff
@@ -292,37 +313,67 @@ async fn main() -> Result<(), Box<dyn Error>> {
     // Video Converter Element
-    let video_converter = gst::ElementFactory::make("videoconvert").build()?;
+    let video_converter = gstreamer::ElementFactory::make("videoconvert").build()?;
 
     // Video Encoder Element
-    let video_encoder = gst::ElementFactory::make(video_encoder_info.name.as_str()).build()?;
+    let video_encoder =
+        gstreamer::ElementFactory::make(video_encoder_info.name.as_str()).build()?;
     video_encoder_info.apply_parameters(&video_encoder, args.app.verbose);
```
Good catch, I'm so used to DMA-BUF working for most GPUs by now I didn't check the other code path properly. Solved in upcoming commit.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -141,6 +212,11 @@ main_loop() {
             log "compositor died."
             increment_retry "compositor"
             start_compositor
+        # Check patcher
+        elif [[ -n "${PATCHER_PID:-}" ]] && ! kill -0 "${PATCHER_PID}" 2>/dev/null; then
+            log "steam-patcher died."
+            increment_retry "steam-patcher"
+            setup_steam_namespaceless
         fi
```
Good catch. Solved in upcoming commit.

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -116,18 +124,50 @@ install_nvidia_driver() {
 }
 
 main() {
+    # Configure and start SSH if enabled
+    if [ "${SSH_ENABLE_PORT}" -ne 0 ]; then
+        log "Configuring SSH server to listen on port ${SSH_ENABLE_PORT}"
+        # Ensure SSH host keys exist
+        ssh-keygen -A 2>/dev/null || {
+            log "Error: Failed to generate SSH host keys."
+            exit 1
+        }
+        # Update SSHD config with the specified port
+        sed -i "s/^#Port .*/Port ${SSH_ENABLE_PORT}/" /etc/ssh/sshd_config || {
+            log "Error: Failed to update SSH port configuration."
+            exit 1
+        }
+        # Start SSH server in background
+        log "Starting SSH server on port ${SSH_ENABLE_PORT}..."
+        /usr/sbin/sshd -D -p "${SSH_ENABLE_PORT}" &
+        SSH_PID=$!
+        log "SSH server started with PID ${SSH_PID}"
+    else
+        log "SSH_ENABLE_PORT set to 0, skipping SSH setup.."
+    fi
```
Fair catch, added in upcoming commit.

DatCaptainHorse committed:
Nitpick fixes

DatCaptainHorse commented on packages/server/src/gpu.rs:
```diff
@@ -162,10 +165,42 @@ pub fn get_gpu_by_card_path(gpus: &[GPUInfo], path: &str) -> Option<GPUInfo> {
-    } else {
-        Some(gpus[index as usize].clone())
+pub fn get_nvidia_gpu_by_cuda_id(gpus: &[GPUInfo], cuda_device_id: usize) -> Option<GPUInfo> {
+    // Check if nvidia-smi is available
+    if Command::new("which").arg("nvidia-smi").output().ok()?.status.success() == false {
+        tracing::warn!("nvidia-smi not found, cannot get NVIDIA GPU by CUDA ID");
+        return None;
+    }
+    
+    // Run nvidia-smi to get information about the CUDA device
+    let output = Command::new("nvidia-smi")
+        .args([
+            "--query-gpu=pci.bus_id",
+            "--format=csv,noheader",
+            "-i",
+            &cuda_device_id.to_string(),
+        ])
+        .output()
+        .ok()?;
+
+    if !output.status.success() {
+        return None;
     }
+
+    // Parse the output to get the PCI bus ID
+    let pci_bus_id = str::from_utf8(&output.stdout).ok()?.trim().to_uppercase(); // nvidia-smi returns uppercase PCI IDs
+
+    // Convert from 00000000:05:00.0 to 05:00.0 if needed
+    let pci_bus_id = if pci_bus_id.starts_with("00000000:") {
+        pci_bus_id[9..].to_string() // Skip the domain part
+    } else if pci_bus_id.starts_with("0000:") {
+        pci_bus_id[5..].to_string() // Alternate check for older nvidia-smi versions
+    } else {
+        pci_bus_id
+    };
+
+    // Find the GPU with the matching PCI bus ID
+    gpus.iter()
+        .find(|gpu| gpu.vendor == GPUVendor::NVIDIA && gpu.pci_bus_id.to_uppercase() == pci_bus_id)
+        .cloned()
```
o\ you yourself in previous nitpick recommended `which`, I had a feeling it would be less-good than just invoking `nvidia-smi` directly, but I trusted your judgement so.. well I'm solving it in upcoming commit.

DatCaptainHorse commented on containers/runner.Containerfile:
```diff
@@ -153,13 +154,10 @@ RUN --mount=type=cache,target=/var/cache/pacman/pkg \
-    tar -xzvf ludusavi.tar.gz && \
-    mv ludusavi /usr/bin/ && \
-    rm ludusavi.tar.gz
+### SSH Configuration ###
+RUN ssh-keygen -A && \
+    echo "PermitRootLogin no" >> /etc/ssh/sshd_config && \
+    echo "PasswordAuthentication yes" >> /etc/ssh/sshd_config
```
Makes sense, solved in upcoming commit.

DatCaptainHorse committed:
More nitpickies

AquaWolf commented on containers/runner.Containerfile:
```diff
@@ -153,13 +154,9 @@ RUN --mount=type=cache,target=/var/cache/pacman/pkg \
+### SSH Configuration ###
+RUN echo "PermitRootLogin no" >> /etc/ssh/sshd_config && \
+    echo "PasswordAuthentication yes" >> /etc/ssh/sshd_config
 
```
I don't like the idea to have standard passwords and password authentication. Please we need to find a better way for ssh access!
Don't merge that.

DatCaptainHorse replied:
It's not enabled by default, and it's meant for debugging only.

DatCaptainHorse replied:
I've added public-key only authentication now, sorry for the delay!
Please resolve and review if you deem the changes good and safe :pray:

DatCaptainHorse replied:
Resolved as per Discord messages.

DatCaptainHorse committed:
SSH with given key only and other niceties

- Allow SSH with only given public key and port
- Update .dockerignore to not build with JetBrains IDE folders (guess who found out they bloated image size..)
- Custom launch application and compositor support
- Allow entrypoint to pass even if it can't change ownership of all home folders (i.e. mounted directory)

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -14,8 +14,8 @@ log() {
     if ! chown -R -h --no-preserve-root "$user_group" "${HOME}" 2>/dev/null; then
-        echo "Error: Failed to change ownership of ${HOME} to ${user_group}" >&2
-        return 1
+        echo "Warning: Failed to change ownership of ${HOME} to ${user_group}" >&2
+        return 0
     fi
     return 0
```
This is a change that was made for a reason, as any mounted volumes (that need to be in /home for Steam to work with them) will fail to change ownership due to the recursive command. Though if you want, I can make it so it's not recursive, only home directory itself will be chown'ed.

DatCaptainHorse replied:
In next commit I'll make it non-recursive and make it fail if it can't chown home directory itself.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -71,33 +132,92 @@ start_nestri_server() {
-    rm -rf /tmp/.X11-unix && mkdir -p /tmp/.X11-unix && chown nestri:nestri /tmp/.X11-unix
-    WAYLAND_DISPLAY=wayland-1 gamescope --backend wayland -g -f -e --rt --mangoapp -W "${WIDTH}" -H "${HEIGHT}" -- steam-native -tenfoot -cef-force-gpu &
-    COMPOSITOR_PID=$!
+    # Set default values only if variables are unset (not empty)
+    if [[ -z "${NESTRI_LAUNCH_CMD+x}" ]]; then
+        NESTRI_LAUNCH_CMD="steam-native -tenfoot -cef-force-gpu"
+    fi
+    if [[ -z "${NESTRI_LAUNCH_COMPOSITOR+x}" ]]; then
+        NESTRI_LAUNCH_COMPOSITOR="gamescope --backend wayland --force-grab-cursor -g -f --rt --mangoapp -W ${WIDTH} -H ${HEIGHT} -r ${FRAMERATE}"
+    fi
 
```
Good catch. Resolving in next commit.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -122,6 +242,9 @@ cleanup() {
     log "Terminating processes..."
     kill_if_running "${NESTRI_PID:-}" "nestri-server"
     kill_if_running "${COMPOSITOR_PID:-}" "compositor"
+    kill_if_running "${APP_PID:-}" "application"
+    kill_if_running "${PATCHER_PID:-}" "steam-patcher"
+    rm -f "/tmp/_v2-entry-point.padded" 2>/dev/null
     exit 0
```
Nice, will do in next commit.

DatCaptainHorse committed:
Nitpicks

DatCaptainHorse committed:
Nitpick missing nestri ssh folder

DatCaptainHorse committed:
Last few nitpicks

DatCaptainHorse commented on packages/scripts/entrypoint.sh:
```diff
@@ -13,7 +13,7 @@ log() {
 # Ensures user directory ownership
 chown_user_directory() {
     local user_group="${USER}:${GID}"
-    if ! chown -R -h --no-preserve-root "$user_group" "${HOME}" 2>/dev/null; then
+    if ! chown -h --no-preserve-root "$user_group" "${HOME}" 2>/dev/null; then
         echo "Error: Failed to change ownership of ${HOME} to ${user_group}" >&2
         return 1
```
Hm, GID is set on the runner Containerfile itself, set to 1000:
```Dockerfile
### User Configuration ###
ENV USER="nestri" \
    UID=1000 \
    GID=1000 \
    USER_PWD="nestri1234" \
    XDG_RUNTIME_DIR=/run/user/1000 \
    HOME=/home/nestri \
    NVIDIA_DRIVER_CAPABILITIES=all
```

So if you could resolve this, please do.

DatCaptainHorse committed:
Last nitpick hopefully

DatCaptainHorse commented:
@coderabbitai Please review, nitpicks hopefully dealt with now.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -71,33 +134,93 @@ start_nestri_server() {
+                log "Compositor socket ready ($COMPOSITOR_SOCKET)."
+                # Patch resolution with wlr-randr for non-gamescope compositors
+                if ! $is_gamescope; then
+                    local OUTPUT_NAME
+                    OUTPUT_NAME=$(WAYLAND_DISPLAY=wayland-0 wlr-randr --json | jq -r '.[] | select(.enabled == true) | .name' | head -n 1)
+                    if [ -z "$OUTPUT_NAME" ]; then
+                        log "Warning: No enabled outputs detected. Skipping wlr-randr resolution patch."
+                        return
+                    fi
+                    WAYLAND_DISPLAY="$COMPOSITOR_SOCKET" wlr-randr --output "$OUTPUT_NAME" --custom-mode "$WIDTH"x"$HEIGHT"
```
Oops, good catch. I'll just do `wayland-0` as that's realistically only wayland socket we'll see. Resolved in next commit.

DatCaptainHorse commented on packages/scripts/entrypoint_nestri.sh:
```diff
@@ -71,33 +134,93 @@
-    rm -rf /tmp/.X11-unix && mkdir -p /tmp/.X11-unix && chown nestri:nestri /tmp/.X11-unix
-    WAYLAND_DISPLAY=wayland-1 gamescope --backend wayland -g -f -e --rt --mangoapp -W "${WIDTH}" -H "${HEIGHT}" -- steam-native -tenfoot -cef-force-gpu &
-    COMPOSITOR_PID=$!
+    # Set default values only if variables are unset (not empty)
+    if [[ -z "${NESTRI_LAUNCH_CMD+x}" ]]; then
+        NESTRI_LAUNCH_CMD="steam-native -tenfoot -cef-force-gpu"
+    fi
+    if [[ -z "${NESTRI_LAUNCH_COMPOSITOR+x}" ]]; then
+        NESTRI_LAUNCH_COMPOSITOR="gamescope --backend wayland --force-grab-cursor -g -f --rt --mangoapp -W ${WIDTH} -H ${HEIGHT} -r ${FRAMERATE:-60}"
+    fi
```
This is intended, to allow no compositor or application to be launched by default on startup.

DatCaptainHorse committed:
Last nitpick hopefully 2

wanjohiryan reviewed (approved):
LGTM
